test/e2e/petset.go

```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
 			pst.verifyPodAtIndex(updateIndex, ps, verify)
 		})
 
+		It("scaling down before scale up is finished should wait until current pod will be running and ready before it will be removed", func() {
+			By("creating petset " + psName + " in namespace " + ns)
+			testProbe := &api.Probe{Handler: api.Handler{HTTPGet: &api.HTTPGetAction{
+				Path: "/index.html",
+				Port: intstr.IntOrString{IntVal: 80}}}}
+			ps := newStatefulSet(psName, ns, headlessSvcName, 1, nil, nil, labels, testProbe)
+			setInitializedAnnotation(ps, "false")
+			ps, err := c.Apps().StatefulSets(ns).Create(ps)
+			Expect(err).NotTo(HaveOccurred())
+			pst := &statefulSetTester{c: c}
+			pst.waitForRunningAndReady(1, ps)
+
+			By("scaling up stateful set " + psName + " to 3 replicas")
+			pst.setHealthy(ps)
+			pst.updateReplicas(ps, 3)
+			pst.waitForRunningAndReady(2, ps)
+
+			By("before scale up finished setting 2nd pod to be not ready by breaking readiness probe")
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+
+			By("scaling down to 1 replica and restoring health of 2nd pod")
+			pst.setHealthy(ps)
+			pst.updateReplicas(ps, 1)
+
+			By("verifying that 2nd pod will not be removed if it is not healthy")
+			pst.confirmPetCount(2, ps, 10*time.Second)
+			expectedPodName := strings.Join([]string{ps.Name, "-1"}, "")
+			expectedPod, err := f.ClientSet.Core().Pods(ns).Get(expectedPodName)
+			Expect(err).NotTo(HaveOccurred())
+			watcher, err := f.ClientSet.Core().Pods(ns).Watch(api.SingleObject(
+				api.ObjectMeta{
+					Name:            expectedPod.Name,
+					ResourceVersion: expectedPod.ResourceVersion,
+				},
+			))
+			Expect(err).NotTo(HaveOccurred())
+			pst.restoreProbe(ps, testProbe)
+
+			By("verifying that 2nd pod will enter running and ready before it will be removed by scale down")
```

Comment:
"Verifying the 2nd pod is not removed by scaling down until it becomes running and ready"